File: data-jpa/src/main/java/com/example/data/jpa/CLR.java
package com.example.data.jpa;

import java.util.Arrays;
import java.util.List;
import java.util.Set;

import com.example.data.jpa.model.Author;
import com.example.data.jpa.model.Book;

import org.springframework.boot.CommandLineRunner;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
class CLR implements CommandLineRunner {

	private final AuthorRepository authorRepository;

	CLR(AuthorRepository authorRepository) {
		this.authorRepository = authorRepository;
	}

	@Override
	@Transactional
	public void run(String... args) {
		var authors = insertAuthors();
		listAllAuthors();
		findById(authors);
		findByPartialName();
		queryFindByName();
		deleteAll();
	}

	private void deleteAll() {
		this.authorRepository.deleteAll();
		long count = this.authorRepository.count();
		System.out.printf("deleteAll(): count = %d%n", count);
	}

	private void queryFindByName() {
		Author author1 = this.authorRepository.queryFindByName("Josh Long").orElse(null);
		Author author2 = this.authorRepository.queryFindByName("Martin Kleppmann").orElse(null);

		System.out.printf("queryFindByName(): author1 = %s%n", author1);
		System.out.printf("queryFindByName(): author2 = %s%n", author2);
	}

	private void findByPartialName() {
		Author author1 = this.authorRepository.findByNameContainingIgnoreCase("sh lo").orElse(null);
		Author author2 = this.authorRepository.findByNameContainingIgnoreCase("in kl").orElse(null);

		System.out.printf("findByPartialName(): author1 = %s%n", author1);
		System.out.printf("findByPartialName(): author2 = %s%n", author2);
	}

	private void findById(List<Author> authors) {
		Author author1 = this.authorRepository.findById(authors.get(0).getId()).orElse(null);
		Author author2 = this.authorRepository.findById(authors.get(1).getId()).orElse(null);

		System.out.printf("findById(): author1 = %s%n", author1);
		System.out.printf("findById(): author2 = %s%n", author2);
	}

	private void listAllAuthors() {
		List<Author> authors = this.authorRepository.findAll();
		for (Author author : authors) {
			System.out.printf("listAllAuthors(): author = %s%n", author);
			for (Book book : author.getBooks()) {
				System.out.printf("\t%s%n", book);
			}
		}
	}

	private List<Author> insertAuthors() {
		Author author1 = this.authorRepository.save(new Author(null, "Josh Long",
				Set.of(new Book(null, "Reactive Spring"), new Book(null, "Cloud Native Java"))));
		Author author2 = this.authorRepository.save(
				new Author(null, "Martin Kleppmann", Set.of(new Book(null, "Designing Data Intensive Applications"))));

		System.out.printf("insertAuthors(): author1 = %s%n", author1);
		System.out.printf("insertAuthors(): author2 = %s%n", author2);

		return Arrays.asList(author1, author2);
	}

}

File: data/data-jdbc-h2-kotlin/src/main/kotlin/com/example/data/jdbc/h2/kotlin/CLR.kt
package com.example.data.jdbc.h2.kotlin

import com.example.data.jdbc.h2.kotlin.model.Author
import com.example.data.jdbc.h2.kotlin.model.Book
import org.springframework.boot.CommandLineRunner
import org.springframework.stereotype.Component

@Component
class CLR(private val authorRepository: AuthorRepository) : CommandLineRunner {
	override fun run(vararg args: String) {
		insertAuthors()
		listAllAuthors()
		findById()
		findByPartialName()
		queryFindByName()
		deleteAll()
	}

	private fun deleteAll() {
		authorRepository.deleteAll()
		val count = authorRepository.count()
		System.out.printf("deleteAll(): count = %d%n", count)
	}

	private fun queryFindByName() {
		val author1 = authorRepository.queryFindByName("Josh Long")
		val author2 = authorRepository.queryFindByName("Martin Kleppmann")
		System.out.printf("queryFindByName(): author1 = %s%n", author1)
		System.out.printf("queryFindByName(): author2 = %s%n", author2)
	}

	private fun findByPartialName() {
		val author1 =
			authorRepository.findByNameContainingIgnoreCase("sh lo")
		val author2 =
			authorRepository.findByNameContainingIgnoreCase("in kl")
		System.out.printf("findByPartialName(): author1 = %s%n", author1)
		System.out.printf("findByPartialName(): author2 = %s%n", author2)
	}

	private fun findById() {
		val author1 = authorRepository.findById(1L).orElse(null)
		val author2 = authorRepository.findById(2L).orElse(null)
		System.out.printf("findById(): author1 = %s%n", author1)
		System.out.printf("findById(): author2 = %s%n", author2)
	}

	private fun listAllAuthors() {
		val authors = authorRepository.findAll()
		for (author in authors) {
			System.out.printf("listAllAuthors(): author = %s%n", author)
			for (book in author.books) {
				System.out.printf("\t%s%n", book)
			}
		}
	}

	private fun insertAuthors() {
		val author1 = authorRepository.save(
			Author(
				null, "Josh Long",
				setOf(Book(null, "Reactive Spring"), Book(null, "Cloud Native Java"))
			)
		)
		val author2 = authorRepository.save(
			Author(
				null,
				"Martin Kleppmann",
				setOf(Book(null, "Designing Data Intensive Applications"))
			)
		)
		System.out.printf("insertAuthors(): author1 = %s%n", author1)
		System.out.printf("insertAuthors(): author2 = %s%n", author2)
	}
}
